### PR TITLE
Enable epochs handler integration part 5

### PR DESCRIPTION
**Pull request by sstanculeanu**
- refactored vm/systemSmartContracts
- added switchHysteresisForMinNodesCurrentEpochFlag, delegationSmartContractCurrentEpochFlag
- completely refactored:
   - StakeEnableEpoch
   - ValidatorToDelegationEnableEpoch
   - DelegationSmartContractEnableEpoch
   - DelegationManagerEnableEpoch
   - ReDelegateBelowMinCheckEnableEpoch
   - ComputeRewardCheckpointEnableEpoch
   - AddTokensToDelegationEnableEpoch
   - DeleteDelegatorAfterClaimRewardsEnableEpoch
   - CorrectLastUnjailedEnableEpoch
   - CorrectFirstQueuedEpoch
   - CorrectJailedNotUnstakedEmptyQueueEpoch
   - SwitchHysteresisForMinNodesEnableEpoch
   - SaveJailedAlwaysEnableEpoch
   - DoubleKeyProtectionEnableEpoch
   - UnbondTokensV2EnableEpoch
   - GovernanceEnableEpoch
   - ESDTEnableEpoch
   - ESDTNFTCreateOnMultiShardEnableEpoch
   - MetaESDTSetEnableEpoch
   - ESDTRegisterAndSetAllRolesEnableEpoch
   
**Notes:**
1. In order to properly check if the old flags match the new methods, one who review should check old conditions from removed `EpochConfirmed` method, the value from config that was passed before, and the line from `EpochConfirmed` from `enableEpochsHandler.go` which uses the same config value.
2. No focus on long tests so far, they will be fixed on the last PR once all components use `enableEpochsHandler`.

### CODE REVIEW

File: common/enableEpochs/enableEpochsHandler.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ func (handler *enableEpochsHandler) EpochConfirmed(epoch uint32, _ uint64) {
 	handler.setFlagValue(epoch >= handler.enableEpochsConfig.SwitchJailWaitingEnableEpoch, handler.switchJailWaitingFlag, "switchJailWaitingFlag")
 	handler.setFlagValue(epoch >= handler.enableEpochsConfig.BelowSignedThresholdEnableEpoch, handler.belowSignedThresholdFlag, "belowSignedThresholdFlag")
 	handler.setFlagValue(epoch >= handler.enableEpochsConfig.SwitchHysteresisForMinNodesEnableEpoch, handler.switchHysteresisForMinNodesFlag, "switchHysteresisForMinNodesFlag")
+	handler.setFlagValue(epoch == handler.enableEpochsConfig.SwitchHysteresisForMinNodesEnableEpoch, handler.switchHysteresisForMinNodesCurrentEpochFlag, "switchHysteresisForMinNodesCurrentEpochFlag")
```
Comment (bogdan-rosianu) [inline common/enableEpochs/enableEpochsHandler.go:43]: might remove the `CurrentEpoch` from the name? also for `delegationSmartContractCurrentEpochFlag`

File: common/enableEpochs/flags.go [hunk 2]
```diff
@@ -3,146 +3,150 @@ package enableEpochs
...
+	aheadOfTimeGasUsageFlag                     *atomic.Flag
+	gasPriceModifierFlag                        *atomic.Flag
+	repairCallbackFlag                          *atomic.Flag
+	balanceWaitingListsFlag                     *atomic.Flag
+	returnDataToLastTransferFlag                *atomic.Flag
+	senderInOutTransferFlag                     *atomic.Flag
+	stakeFlag                                   *atomic.Flag
+	stakingV2Flag                               *atomic.Flag
+	stakingV2OwnerFlag                          *atomic.Flag
+	stakingV2GreaterEpochFlag                     *atomic.Flag
```
Comment (bogdan-rosianu) [inline common/enableEpochs/flags.go:25]: code not alligned

File: common/enableEpochs/flags.go [hunk 3]
```diff
@@ -3,146 +3,150 @@ package enableEpochs
...
+		aheadOfTimeGasUsageFlag:                     &atomic.Flag{},
+		gasPriceModifierFlag:                        &atomic.Flag{},
+		repairCallbackFlag:                          &atomic.Flag{},
+		balanceWaitingListsFlag:                     &atomic.Flag{},
+		returnDataToLastTransferFlag:                &atomic.Flag{},
+		senderInOutTransferFlag:                     &atomic.Flag{},
+		stakeFlag:                                   &atomic.Flag{},
+		stakingV2Flag:                               &atomic.Flag{},
+		stakingV2OwnerFlag:                          &atomic.Flag{},
+		stakingV2GreaterEpochFlag:                     &atomic.Flag{},
```
Comment (bogdan-rosianu) [inline common/enableEpochs/flags.go:99]: not alligned

File: common/enableEpochs/flags.go [hunk 4]
```diff
@@ -186,6 +190,11 @@ func (fh *flagsHolder) IsSwitchHysteresisForMinNodesFlagEnabled() bool {
 	return fh.switchHysteresisForMinNodesFlag.IsSet()
 }
 
+// IsSwitchHysteresisForMinNodesFlagEnabledForCurrentEpoch returns true if switchHysteresisForMinNodesCurrentEpochFlag is enabled
+func (fh *flagsHolder) IsSwitchHysteresisForMinNodesFlagEnabledForCurrentEpoch() bool {
```
Comment (bogdan-rosianu) [inline common/enableEpochs/flags.go:194]: remove `ForCurrentEpoch` from the name?

File: common/enableEpochs/flags.go [hunk 5]
```diff
@@ -281,6 +290,11 @@ func (fh *flagsHolder) IsDelegationSmartContractFlagEnabled() bool {
 	return fh.delegationSmartContractFlag.IsSet()
 }
 
+// IsDelegationSmartContractFlagEnabledForCurrentEpoch returns true if delegationSmartContractCurrentEpochFlag is enabled
+func (fh *flagsHolder) IsDelegationSmartContractFlagEnabledForCurrentEpoch() bool {
```
Comment (bogdan-rosianu) [inline common/enableEpochs/flags.go:294]: remove `ForCurrentEpoch` from name?

File: epochStart/metachain/systemSCs.go [hunk 6]
```diff
@@ -284,15 +254,15 @@ func (s *systemSCProcessor) ProcessSystemSmartContract(
 		}
 	}
 
-	if s.flagESDTEnabled.IsSet() {
+	if s.enableEpochsHandler.IsESDTFlagEnabledForCurrentEpoch() {
```
Comment (bogdan-rosianu) [inline epochStart/metachain/systemSCs.go:257]: remove the `CurrentEpoch` from the naming? applicable for all changes, so I won't leave too many comments

File: vm/systemSmartContracts/delegation_test.go [hunk 7]
```diff
@@ -4476,19 +4479,20 @@ func TestDelegation_checkArgumentsForValidatorToDelegation(t *testing.T) {
 			return vmcommon.Ok
 		}}, nil
 	}})
+	stub, _ := args.EnableEpochsHandler.(*testscommon.EnableEpochsHandlerStub)
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/delegation_test.go:4482]: epochsHandler instead of stub

File: vm/systemSmartContracts/esdt.go [hunk 8]
```diff
@@ -1463,10 +1432,6 @@ func getFirstAddressWithGivenRole(token *ESDTDataV2, definedRole []byte) ([]byte
 }
 
 func (e *esdt) changeToMultiShardCreate(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	if !e.flagTransformToMultiShardCreate.IsSet() {
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/esdt.go:1466]: why did you remove this?

File: vm/systemSmartContracts/staking_test.go [hunk 9]
```diff
@@ -2261,8 +2255,8 @@ func TestStakingSc_UnStakeNodeWhenMaxNumIsMoreShouldNotStakeFromWaiting(t *testi
 	args.StakingSCConfig.MinStakeValue = stakeValue.Text(10)
 	args.StakingSCConfig.MaxNumberOfNodesForStake = 2
 	args.MinNumNodes = 1
-	args.EpochConfig.EnableEpochs.StakingV2EnableEpoch = 0
-	args.EpochConfig.EnableEpochs.CorrectLastUnjailedEnableEpoch = 0
+	stub, _ := args.EnableEpochsHandler.(*testscommon.EnableEpochsHandlerStub)
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/staking_test.go:2258]: epochsHandler instead of stub

File: vm/systemSmartContracts/validator_test.go [hunk 10]
```diff
@@ -425,7 +424,8 @@ func TestStakingValidatorSC_ExecuteStakeDoubleKeyAndCleanup(t *testing.T) {
 
 	args.Eei = eei
 	args.StakingSCConfig = argsStaking.StakingSCConfig
-	args.EpochConfig.EnableEpochs.DoubleKeyProtectionEnableEpoch = 1000
+	stub, _ := args.EnableEpochsHandler.(*testscommon.EnableEpochsHandlerStub)
```
Comment (bogdan-rosianu) [inline vm/systemSmartContracts/validator_test.go:427]: stub -> epochsHandler

File: common/enableEpochs/enableEpochsHandler.go (hunk 1 above)
Comment (sstanculeanu) [reply common/enableEpochs/enableEpochsHandler.go:43]: not needed, it would be the same with the one from L42..`CurrentEpoch` suggests that it is set on `epoch == epoch_from_config`..same for others, I'll just mark them as resolved

File: vm/systemSmartContracts/delegation_test.go (hunk 7 above)
Comment (sstanculeanu) [reply vm/systemSmartContracts/delegation_test.go:4482]: replaced in all places

File: vm/systemSmartContracts/esdt.go (hunk 8 above)
Comment (sstanculeanu) [reply vm/systemSmartContracts/esdt.go:1466]: this flag was not set anywhere

File: epochStart/metachain/systemSCs_test.go [hunk 11]
```diff
@@ -1385,7 +1398,9 @@ func TestSystemSCProcessor_ProcessSystemSmartContractUnStakeOneNodeStakeOthers(t
 		_ = args.PeerAccountsDB.SaveAccount(jailedAcc)
 	}
 
-	s.flagSetOwnerEnabled.Reset()
+	args.EpochNotifier.CheckEpoch(&testscommon.HeaderHandlerStub{
+		EpochField: 1, // disable stakingV2OwnerFlag
+	})
 	err := s.ProcessSystemSmartContract(validatorInfos, 0, 0)
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply epochStart/metachain/systemSCs_test.go:1404]: at the beginning of the test, `createFullArgumentsForSystemSCProcessing` gets an empty config, which means all flags are set to 0. For this specific tests, we want to bypass [this check](https://github.com/ElrondNetwork/elrond-go/blob/integrate_enable_epochs_handler_p5/epochStart/metachain/systemSCs.go#L188). `IsStakingV2OwnerFlagEnabled` returns true if current epoch == staking v2 epoch. As we use real components, calling `CheckEpoch` with 1 will make `IsStakingV2OwnerFlagEnabled` return false, but `IsStakingV2FlagEnabled` return true

File: vm/systemSmartContracts/esdt_test.go [hunk 12]
```diff
@@ -4370,92 +4382,6 @@ func TestEsdt_ExecuteIssueSFTAndChangeSFTToMetaESDT(t *testing.T) {
 	assert.True(t, strings.Contains(eei.returnMessage, "change can happen to semi fungible tokens only"))
 }
 
-func TestEsdt_ExecuteTransformToMultiShardCreate(t *testing.T) {
-	t.Parallel()
-
-	args := createMockArgumentsForESDT()
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply vm/systemSmartContracts/esdt_test.go:4376]: from what I've seen, this test was based on the flag set using `transformToMultiShardCreateEnableEpoch`, which was not set/used at all in the component.. As the flag was completely removed, the test was removed as well.

**Review by ssd04 — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]
